### PR TITLE
Add a DeletePrefix method to support efficient recursive deletes of l…

### PR DESCRIPTION
**Pull request by preetapan**
…arge subtrees, and unit tests

### CODE REVIEW

File: iradix.go [hunk 1]
```diff
@@ -183,6 +183,32 @@ func (t *Txn) writeNode(n *Node, forLeafUpdate bool) *Node {
...
+	if t.trackMutate {
+		t.trackChannel(n.mutateCh)
+	}
+
+	// Mark its leaf as being mutated, if appropriate.
+	if t.trackMutate && n.leaf != nil {
+		t.trackChannel(n.leaf.mutateCh)
+	}
+
+	if n.leaf != nil && fn(n) {
```
Comment (slackpad) [inline iradix.go]: Why do we check the callback's return value here? That always returns `false` so this counts as a side effect (for the leaf) but exits in kind of an obscure way. I think we should get rid of the callback and just have this keep track of the count and return it; that'll simplify things and make the behavior super clear, since we don't have any other callback behavior to hook into here.

File: iradix.go [hunk 2]
```diff
@@ -357,6 +383,61 @@ func (t *Txn) delete(parent, n *Node, search []byte) (*Node, *leafNode) {
 	return nc, leaf
 }
 
+// delete does a recursive deletion
+func (t *Txn) deletePrefix(parent, n *Node, search []byte) (*Node, int) {
+	// Check for key exhaustion
+	if len(search) == 0 {
+		// Remove the leaf node
+		nc := t.writeNode(n, true)
+		subTreeSize := 0 //count this node
```
Comment (slackpad) [inline iradix.go]: This is confusing that you say "count this node" but set the count to zero?

File: iradix.go [hunk 3]
```diff
@@ -357,6 +383,61 @@ func (t *Txn) delete(parent, n *Node, search []byte) (*Node, *leafNode) {
...
+		}
+		nc.edges = nil // deletes the entire subtree
+
+		return nc, subTreeSize
+	}
+
+	// Look for an edge
+	label := search[0]
+	idx, child := n.getEdge(label)
+	if child == nil || (!bytes.HasPrefix(child.prefix, search) && !bytes.HasPrefix(search, child.prefix)) {
```
Comment (slackpad) [inline iradix.go]: How come you needed to compare both ways here?

File: iradix.go (hunk 1 above)
Comment (preetapan) [inline iradix.go]: I started this modelling some other walk functions I saw in the code base, agree now that this is too confusing and doesn't need to be implemented this way. I'll simplify it

File: iradix.go (hunk 3 above)
Comment (preetapan) [inline iradix.go]: The first condition handles the case when the prefix being deleted does not have an associated leaf node. E.g when you want to delete the prefix `"AB"`, in a tree that contains `"ABCDE"` and `"ABCD"`, then at this point child.prefix will be `"ABCDE"` and will contain `"AB"` as its prefix.  

I'll add some more test cases, i want to make sure i didn't miss other edge cases here.

File: iradix.go [hunk 4]
```diff
@@ -183,6 +183,28 @@ func (t *Txn) writeNode(n *Node, forLeafUpdate bool) *Node {
 	return nc
 }
 
+// Visit all the nodes in the tree under n, and add their mutateChannels to the transaction
+// Returns the size of the subtree visited
+func (t *Txn) recursiveTrackChannelAndCount(n *Node) int {
```
Comment (preetapan) [inline iradix.go]: Much simpler method that both counts the nodes visited and tracks mutate channels correctly

File: iradix.go (hunk 3 above)
Comment (slackpad) [inline iradix.go]: A tree with "ab", "acats", and "acato" added might be good:

```
     [root]
       | a
     [node] (prefix=a)
    b/    \c
   (ab)  [node] (prefix=cat)
        s/    \o
     (acats) (acato)
```

I think if we can run cases where we delete prefix "a", "ab", "ac", "aca", "acat", "acats", "acato" we will get the edge cases of leaves, nodes with leaves under them, nodes with nodes and leaves, and deleting a prefix that lands on a node with a longer prefix.

Comment (slackpad) [inline iradix.go]: I think I'm finally understanding this a bit, though still going through some cases but i think the same prefix handling logic from `delete` should be sufficient, just the `len(search) == 0` case is different, as you've done.

```
         // Look for an edge
         label := search[0]
         idx, child := n.getEdge(label)
         if child == nil || !bytes.HasPrefix(search, child.prefix) {
                 return nil, nil
         }

         // Consume the search prefix
         search = search[len(child.prefix):]
         newChild, leaf := t.delete(n, child, search)
         if newChild == nil {
                 return nil, nil
         }
```

For the case you were worried about, if you delete "aca" prefix it will land on the node with the prefix "cat", consume "ca", and then call again with the empty search prefix, which will do the right thing. I don't think we could get down here for the other cases we are checking for in the PR, because we have to have the right prefix leading up to the node we are on. The only case that can happen is that the node has more stuff than our prefix, but that exhausts the search and terminates the algorithm.

Comment (slackpad) [inline iradix.go]: Alright I think I understand why we do both lookups now. I'd set the search to `""` in the case where it's shorter, just to be clearer. Would also be good to add a comment talking about what case we are trying to catch here.

File: iradix.go (hunk 4 above)
Comment (slackpad) [inline iradix.go]: I like this form - I'd rename this to `trackChannelsAndCount` to have a little simpler name.

File: iradix.go [hunk 5]
```diff
@@ -357,6 +379,60 @@ func (t *Txn) delete(parent, n *Node, search []byte) (*Node, *leafNode) {
...
+// delete does a recursive deletion
+func (t *Txn) deletePrefix(parent, n *Node, search []byte) (*Node, int) {
+	// Check for key exhaustion
+	if len(search) == 0 {
+		// Remove the leaf node
+		nc := t.writeNode(n, true)
+		if n.isLeaf() {
+			nc.leaf = nil
+		}
+		subTreeSize := 0
```
Comment (slackpad) [inline iradix.go]: ^ this one i think is still valid

File: iradix.go [hunk 6]
```diff
@@ -357,6 +379,60 @@ func (t *Txn) delete(parent, n *Node, search []byte) (*Node, *leafNode) {
...
+	if len(child.prefix) > len(search) {
+		search = search[len(search):]
+	} else {
+		search = search[len(child.prefix):]
+	}
+	newChild, numDeletions := t.deletePrefix(n, child, search)
+	if newChild == nil {
+		return nil, 0
+	}
+	//copy this node
```
Comment (slackpad) [inline iradix.go]: Can you copy the "WATCH OUT" comment here, too, since it still applies.

File: iradix.go [hunk 7]
```diff
@@ -384,6 +460,20 @@ func (t *Txn) Delete(k []byte) (interface{}, bool) {
...
 }
 
+// DeletePrefix is used to delete an entire subtree that matches the prefix
+// This will delete all nodes under that prefix
+func (t *Txn) DeletePrefix(prefix []byte) bool {
+	newRoot, numDeletions := t.deletePrefix(nil, t.root, prefix)
+	if newRoot != nil {
+		t.root = newRoot
+	}
+	if numDeletions > 0 {
```
Comment (slackpad) [inline iradix.go]: This doesn't seem like it can be true unlees `newRoot != nil`, can this just be:

```
if newRoot != nil {
    t.root = newRoot
   t.size = t.size - numDeletions
}
```

File: iradix.go [hunk 8]
```diff
@@ -524,6 +614,14 @@ func (t *Tree) Delete(k []byte) (*Tree, interface{}, bool) {
 	return txn.Commit(), old, ok
 }
 
+// Delete is used to delete a given key. Returns the new tree,
```
Comment (slackpad) [inline iradix.go]: This comment seems stale.

File: iradix_test.go [hunk 9]
```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
...
+	//various test cases where DeletePrefix should succeed
+	cases := []exp{
+		{[]string{"", "test/test1", "test/test2", "test/test3", "R", "RA"}, "test", []string{"", "R", "RA"}},
+		{[]string{"", "test", "test/test1", "test/test2", "test/test3", "test/testAAA", "R", "RA"}, "test", []string{"", "R", "RA"}},
+		{[]string{"", "test/test1", "test/test2", "test/test3", "test/testAAA", "R", "RA"}, "test/test", []string{"", "R", "RA"}},
+		{[]string{"", "AB", "ABC", "AR", "R", "RA"}, "AR", []string{"", "AB", "ABC", "R", "RA"}},
+	}
+
+	for _, testCase := range cases {
+		verifyPrefixDelete(t, testCase.treeNodes, testCase.prefix, testCase.expectedOut)
```
Comment (slackpad) [inline iradix_test.go]: It would be good to add a short description for what each case is trying to hit, then you could do `t.Run()` to run as a sub-test and just inline the `verifyPrefixDelete` code. You could also delete an invalid prefix like "nope" as part of each case and verify it, otherwise I'd factor that out as a separate test entirely.

File: iradix_test.go [hunk 10]
```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
...
+	}
+
+	//verify deleting a prefix that doesn't exist
+	verifyInvalidPrefix(t, []string{"", "test/test1", "test/test2"}, "XXXX")
```
Comment (slackpad) [inline iradix_test.go]: It's a little weird this is on the end here. You could verify deleting a "nope" prefix for each of the test cases and re-verify it matches the expected, so this becomes part of each test case.

File: iradix_test.go [hunk 11]
```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
...
+	default:
+	}
+
+}
+
+func verifyTree(t *testing.T, expected []string, r *Tree) {
+	root := r.Root()
+	out := []string{}
+	fn := func(k []byte, v interface{}) bool {
+		out = append(out, string(k))
```
Comment (slackpad) [inline iradix_test.go]: I'd append like `fmt.Sprintf("|%s", string(k))` so that you can't ever get into trouble with common prefixes running together.

File: iradix_test.go [hunk 12]
```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
...
+	if !ok {
+		t.Fatalf("Expected delete prefix to return true")
+	}
+	if hasAnyClosedMutateCh(r) {
+		t.Fatalf("Transaction was not committed, no channel should have been closed")
+	}
+
+	txn.Commit()
+
+	// Verify root and parent triggered, not leaf affected
```
Comment (slackpad) [inline iradix_test.go]: This comment is stale about the "not leaf affected" part

File: iradix_test.go [hunk 13]
```diff
@@ -212,6 +211,145 @@ func TestDelete(t *testing.T) {
...
+	if r.Len() != len(keys) {
+		t.Fatalf("bad len: %v %v", r.Len(), len(keys))
+	}
+
+	rootWatch, _, _ := r.Root().GetWatch(nil)
+	if rootWatch == nil {
+		t.Fatalf("Should have returned a watch")
+	}
+
+	nodeWatch, _, _ := r.Root().GetWatch([]byte("foo/bar/baz"))
```
Comment (slackpad) [inline iradix_test.go]: For this it might be nice to get all three of the foo ones to make sure we recurse and get all of them.

File: iradix.go (hunk 5 above)
Comment (preetapan) [inline iradix.go]: Addressed this, had to make one minor tweak to `trackChannelsAndCount` to prevent overcounting non-leaf nodes.

**Comment by preetapan** [comment]
I've addressed all the comments. @slackpad please take another last look and then we can merge to master
